internal/auth/permissions.go

```diff
@@ -0,0 +1,106 @@
+package auth
+
+import (
+	"context"
+	"net/http"
+	"strings"
+
+	"github.com/sirupsen/logrus"
+)
+
+const (
+	Forbidden                      = "Forbidden"
+	AuthorizationServerUnavailable = "Authorization server unavailable"
+	BadRequest                     = "Unable to verify request"
+)
+
+func CreatePermissionsMiddleware(log logrus.FieldLogger) func(http.Handler) http.Handler {
+	return func(next http.Handler) http.Handler {
+		fn := func(w http.ResponseWriter, r *http.Request) {
+			parts := strings.Split(r.URL.Path, "/")
+			// /, /api, /api/v{version} and /api/v{version}/auth don't require permissions
+			if len(parts) <= 3 || (len(parts) >= 4 && parts[3] == "auth") {
+				next.ServeHTTP(w, r)
+				return
+			}
+
+			// Extract resource and action from the request
+			// Skip /api/v{version}
+			resource, action := extractResourceAndAction(parts[3:], r.Method)
+			if resource == "" || action == "" {
+				log.Errorf("Unable to extract resource and action from %s and %s", r.URL.Path, r.Method)
+				http.Error(w, BadRequest, http.StatusBadRequest)
+				return
+			}
+
+			if !isAllowed(r.Context(), resource, action, w, log) {
+				// http.Error was called in isAllowed
+				return
+			}
+
+			// If authorized, proceed to the next handler
+			next.ServeHTTP(w, r)
+		}
+		return http.HandlerFunc(fn)
+	}
+}
+
+func isAllowed(ctx context.Context, resource, action string, w http.ResponseWriter, log logrus.FieldLogger) bool {
+	// Perform permission check
+	allowed, err := GetAuthZ().CheckPermission(ctx, resource, action)
+	if err != nil {
+		http.Error(w, AuthorizationServerUnavailable, http.StatusServiceUnavailable)
+		return false
+	}
+	if !allowed {
+		http.Error(w, Forbidden, http.StatusForbidden)
+		return false
+	}
+
+	return true
+}
+
+func extractResourceAndAction(parts []string, method string) (string, string) {
+	// Handle according to the URL structure
+	// e.g., "device", "devices", "devices/{name}", "devices/{name}/sub-action"
+	resource := strings.ToLower(parts[0])
+	action := ""
+	switch method {
+	case http.MethodGet:
+		action = "get"
```

Comment:
Could we make these actions consts to reduce scope of future typos?

Reply:
done